HDDS-11476. Implement lesser/greater operation for --filter option of ldb scan command

Tejaskriya opened the pull request:
## What changes were proposed in this pull request?

A --filter option was added recently which selects records that have a given value for a particular field will make debugging easier. (as a part of https://github.com/apache/ozone/pull/7167)
Adding support for lesser/greater than operations will be useful while debugging
For example, if a value has many fields like [name, location->[address, DN, IP], version, lastUpdateTime], 
- using the option "--filter={version:equals:1}" will display records that have the value 1 for version.
- using the option "–filter={lastUpdateTime:greater:1000}" will display the record with lastUpdateTime>1000
- using the option "–filter={lastUpdateTime:lesser:1000}" will display the record with lastUpdateTime<1000

A list of fields along with the value it should be compared to is given to the command, and only those records passing the condition it will be shown.

eg.) ozone debug ldb --db=/data/metadata/om.db scan --cf=volumeTable --filter="usedNamespace:greater:2,adminName:equals:impala" 

In this PR, apart from adding the support for these 2 operations, some code refactoring is also done.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11476

## How was this patch tested?

Added tests in TestLDBCli
Tested locally with a local DB:
```
$ ozone debug ldb --db=/Users/tejaskriya.madhan/ZDump/bigDB/omMBdb/om.db scan --cf=keyTable  -l=3  --filter=dataSize:greater:2 --fields="objectID,updateID,volumeName,bucketName,keyName,dataSize"

{ "/55406660-10tb/ozone-legacy-ec-1st-bucket-55406660/10T-1-terasort-input/part-m-00000": {
  "updateID" : 2363922,
  "bucketName" : "ozone-legacy-ec-1st-bucket-55406660",
  "volumeName" : "55406660-10tb",
  "keyName" : "10T-1-terasort-input/part-m-00000",
  "dataSize" : 108695652200,
  "objectID" : -9223372036253037824
}
, "/55406660-10tb/ozone-legacy-ec-1st-bucket-55406660/10T-1-terasort-input/part-m-00001": {
  "updateID" : 2363776,
  "bucketName" : "ozone-legacy-ec-1st-bucket-55406660",
  "volumeName" : "55406660-10tb",
  "keyName" : "10T-1-terasort-input/part-m-00001",
  "dataSize" : 108695652200,
  "objectID" : -9223372036253027072
}
, "/55406660-10tb/ozone-legacy-ec-1st-bucket-55406660/10T-1-terasort-input/part-m-00002": {
  "updateID" : 2363418,
  "bucketName" : "ozone-legacy-ec-1st-bucket-55406660",
  "volumeName" : "55406660-10tb",
  "keyName" : "10T-1-terasort-input/part-m-00002",
  "dataSize" : 108695652200,
  "objectID" : -9223372036253021440
}
 }


$ ozone debug ldb --db=/Users/tejaskriya.madhan/ZDump/bigDB/omMBdb/om.db scan --cf=keyTable  -l=3  --filter=dataSize:lesser:2 --fields="objectID,updateID,volumeName,bucketName,keyName,dataSize"

{ "/55406660-10tb/ozone-legacy-ec-1st-bucket-55406660/10T-1-terasort-input/": {
  "updateID" : 2350475,
  "bucketName" : "ozone-legacy-ec-1st-bucket-55406660",
  "volumeName" : "55406660-10tb",
  "keyName" : "10T-1-terasort-input/",
  "dataSize" : 0,
  "objectID" : -9223372036253054206
}
, "/55406660-10tb/ozone-legacy-ec-1st-bucket-55406660/10T-1-terasort-input/_SUCCESS": {
  "updateID" : 2364036,
  "bucketName" : "ozone-legacy-ec-1st-bucket-55406660",
  "volumeName" : "55406660-10tb",
  "keyName" : "10T-1-terasort-input/_SUCCESS",
  "dataSize" : 0,
  "objectID" : -9223372036249583104
}
, "/55406660-10tb/ozone-legacy-ec-1st-bucket-55406660/10T-1-terasort-output/": {
  "updateID" : 2365898,
  "bucketName" : "ozone-legacy-ec-1st-bucket-55406660",
  "volumeName" : "55406660-10tb",
  "keyName" : "10T-1-terasort-output/",
  "dataSize" : 0,
  "objectID" : -9223372036249105919
}
 }


$ ozone debug ldb --db=/Users/tejaskriya.madhan/ZDump/bigDB/omMBdb/om.db scan --cf=keyTable  -l=3  --filter=keyName:lesser:2 --fields="objectID,updateID,volumeName,bucketName,keyName,dataSize"

LESSER or GREATER operation can be performed only on numeric values.
ERROR: field: keyName=(LESSER,2,null), ex: java.io.IOException: Invalid filter passed
{  }
Exit code is non-zero. Check the error message above
```

ayushtkn commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:
```diff
@@ -289,33 +289,60 @@ private void processRecords(ManagedRocksIterator iterator,
+        } else {
+          Filter filterValue = new Filter(fieldValue[1], fieldValue[2]);
+          if (filterValue.getOperator() == null) {
+            err().println("Error: Invalid format for filter \"" + filterValue
```
This isn't invalid `format` for filter, but invalid `operator` for the filter

ayushtkn commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:
```diff
@@ -289,33 +289,60 @@ private void processRecords(ManagedRocksIterator iterator,
+
+    Map<String, Filter> fieldsFilterSplitMap = new HashMap<>();
+    if (filter != null) {
+      for (String field : filter.split(",")) {
```
what is being split by ``,`` here, your filter would be like ``dataSize:lesser:500`` and you can't specify multiple filters by this code if you do so something like ``dataSize:lesser:500, dataSize:equal:500`` this crashes.

Can you help what is this doing? & maybe extend a test case which uses this, if not already

I removed this for loop itself & no test fails in ``TestLDBCLI``, I guess the intention might be to support multiple filters maybe, if so you need to fix it

Tejaskriya replied:
The aim of having this loop is to pass multiple filters like `dataSize:lesser:500, bucketName:equal:bucket1`. It is to apply filters to different fields, not on the same field.
I will add some more tests to cover the scenario.

ayushtkn commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:
```diff
@@ -328,6 +355,152 @@ private void processRecords(ManagedRocksIterator iterator,
+  }
+
+  private boolean checkFilteredObject(Object obj, Class<?> clazz, Map<String, Filter> fieldsSplitMap)
+      throws IOException {
```
this doesn't throw IOE

ayushtkn commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:
```diff
@@ -328,6 +355,152 @@ private void processRecords(ManagedRocksIterator iterator,
+      }
+    }
+    if (classField == null) {
+      throw new NoSuchFieldException();
```
I think we should put some message as well, which can tell which field not found in which class or so

ayushtkn commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:
```diff
@@ -328,6 +355,152 @@ private void processRecords(ManagedRocksIterator iterator,
+        exception = true;
+        return false;
+      } catch (IllegalAccessException e) {
+        err().println("ERROR: Cannot get field from object: " + field);
```
this ain't very right, this communicates as if the object: name is given by field

ayushtkn commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/debug/TestLDBCli.java:
```diff
@@ -182,6 +182,30 @@ private static Stream<Arguments> scanTestCases() {
+            Named.of("Default", Pair.of(0, "")),
+            Named.of("Filter dataSize>2000", Arrays.asList("--filter", "dataSize:greater:2000")),
+            Named.of("Expect empty result", null)
+        ),
```
should add some -ve test cases somewhere as well

Tejaskriya replied:
I have added some negative test cases in the robot test file now

Tejaskriya commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/debug/DBScanner.java:
```diff
@@ -132,7 +132,7 @@ public class DBScanner implements Callable<Void>, SubcommandWithParent {
       description = "Comma-separated list of \"<field>:<operator>:<value>\" where " +
           "<field> is any valid field of the record, " +
-          "<operator> is (EQUALS,MAX or MIN) and " +
+          "<operator> is (EQUALS,LESSER or GREATER) and " +
```
In order to implement MAX and MIN, some code refactoring would be required. We would not be able to do batch processing as max/min cannot be calculated in batches. This could severely affect the performance. So having lesser/greater than operators could perform as a substitute for min/max.

sumitagrawl replied:
Since MIN/MAX operation do not match doing parallel processing of records in batches and generating output, so dropping this. But having alternative with < , >  can be tuned to provide similar functionality of matching.

Tejaskriya commented:
@ayushtkn thank you for the review. I have addressed the comments. Could you please review the updated patch?
@sumitagrawl could you please review my patch when you have some time?

sumitagrawl reviewed (approved):
@Tejaskriya LGTM
